java/flight/flight-core/src/main/java/org/apache/arrow/flight/SessionOptionValue.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+import org.apache.arrow.flight.impl.Flight;
+
+/**
+ * A union-like container interface for supported session option value types.
+ */
+public abstract class SessionOptionValue {
+    static SessionOptionValue fromProto(Flight.SessionOptionValue proto); {
+        switch(proto.getOptionValueCase()) {
+            case STRING_VALUE:
+                return new SessionOptionValueString(proto.getStringValue());
+                break;
+            case BOOL_VALUE:
+                return new SessionOptionValueBoolean(proto.getValue());
+                break;
+            case INT32_VALUE:
+                return new SessionOptionValueInt(proto.getInt32Value());
+                break;
+            case INT64_VALUE:
+                return new SessionOptionValueLong(proto.getInt64Value());
+                break;
+            case FLOAT_VALUE:
+                return new SessionOptionValueFloat(proto.getFloatValue());
+                break;
+            case DOUBLE_VALUE:
+                return new SessionOptionValueDouble(proto.getDoubleValue());
+                break;
+            case STRING_LIST_VALUE:
+                // FIXME PHOXME is this what's in the ProtocolStringList?
+                return new SessionOptionValueStringList(proto.getValueStringList().stream().collect(
+                    Collectors.toList(e -> google.protocol.StringValue.parseFrom(e).getValue())));
+                break;
+            default:
+                // Unreachable
+                 throw new IllegalArgumentException("");
+        }
+    }
+
+    /**
+     * Value access via a caller-provided visitor/functor.
+     */
+    abstract void visit(SessionOptionValueVisitor);
```

Comment:
This should be "acceptVisitor".

Reply:
SessionOptionValueVisitor should probably be a generic and acceptVisitor() should return its result.

Reply:
The callback is required to avoid the called to acceptVisitor also having to be generic (in order to assign a templated type), no?

Reply:
> SessionOptionValueVisitor should probably be a generic and acceptVisitor() should return its result.

I didn't because Java can't constrain the templated type to a specific set of values; SessionOptionValue could possibly—I think that's checked at runtime so the SessionOptionValueVisitor::visit overloads would constrain that correctly?

Reply:
Also the SessionOptionValue<String[]> can't be constructed with the same generic code as all of the other variants.

Reply:
tl;dr I *think* we can make SessionOptionValue generic just fine, using Java.lang.Integer and friends.

Reply:
OTOH Java still doesn't really support specialization to an array type (that I know of?) so I'm not sure we can build this on top of Java generics.  All ears if there's a way to make Java do the thing.

Reply:
On further thought I think there's a way to, er, mostly use generics here—will flesh that out in a bit and see if it adds up how I think.

Reply:
Done, thanks.